ocdata/core/slab.py

```diff
@@ -71,28 +71,31 @@ def __init__(
 
     @classmethod
     def from_bulk_get_random_slab(
-        cls, bulk=None, max_miller=2, tile_and_tag=True, min_ab=8.0
+        cls, bulk=None, max_miller=2, tile_and_tag=True, min_ab=8.0, save_path=None
     ):
-        assert bulk is not None
-
-        slabs = compute_slabs(
-            bulk.atoms,
-            max_miller=max_miller,
+        all_slabs = Slab.from_bulk_get_all_slabs(
+            bulk, max_miller, tile_and_tag, min_ab, save_path
         )
-        slab_idx = np.random.randint(len(slabs))
-        unit_slab_struct, millers, shift, top = slabs[slab_idx]
-        return cls(bulk, unit_slab_struct, millers, shift, top, tile_and_tag, min_ab)
+        slab_idx = np.random.randint(len(all_slabs))
+        return all_slabs[slab_idx]
 
     @classmethod
     def from_bulk_get_all_slabs(
-        cls, bulk=None, max_miller=2, tile_and_tag=True, min_ab=8.0
+        cls, bulk=None, max_miller=2, tile_and_tag=True, min_ab=8.0, save_path=None
     ):
         assert bulk is not None
 
         slabs = compute_slabs(
             bulk.atoms,
             max_miller=max_miller,
         )
+        # if path is provided, save out the pkl
+        if save_path is not None:
+            assert not os.path.exists(save_path)
+            os.makedirs(os.path.dirname(save_path), exist_ok=True)
+            with open(save_path, "wb") as f:
+                pickle.dump(slabs, f)
+
```

Comment:
- Since we are computing the tiled and tagged versions of all of these slabs, would it make sense to save those instead of the intermediate unit slabs?
- More of a style nit: thoughts on moving this saving functionality outside of the slab loading classmethods?

Reply:
I'm not opposed to moving the saving functionality outside of the slab class, but I think this is fine here since it's only a few lines and nothing else would use it (besides adslab metadata, but adslabs have other stuff to write).

Btw, when we generate and write slabs in parallel with structure_generator.py, there's a chance that the same bulk will be processed in two threads simultaneously, meaning that they would both call this method instead of `from_precomputed_slabs_pkl` and write `bulk_id.pkl`. This is not the most efficient but I think it's okay, if anyone has a better method let me know.

Reply:
Hmm I see. Agreed that I don't think it's a huge concern. If anything we could modify it upstream in how we batch things for parallelizing to ensure this, happy with how it currently stands though.

Reply:
This actually results in a hard error - two threads that pick up an identical surface will hit an error when they both try writing surface VASP inputs. Overwriting duplicates wouldn't be an issue, but since this kills the job it is worth addressing.